### ISSUE TITLE
Reduce number of constraints for JubJub IsOnCurve test

### ISSUE THREAD

**Issue by HarryR** [HarryR/ethsnarks#71]
As per @swasilyev insight on https://github.com/HarryR/ethsnarks/pull/70#issuecomment-436347492

```
a*x*x + y*y = 1 + d*x*x*y*y
```

This can be simplified to:

```
// axx + yy = 1 + dxxy*y
(x, x, xx)
(y, y, yy)
(d * xx, yy, a * xx + yy - 1)
```

**Comment by swasilyev** [comment]
It's not my insight. Section A.3.3 of [Zcash Protocol Specification](https://github.com/zcash/zips/blob/master/protocol/sapling.pdf) describes circuits for original (Zcash) jubjub arithmetic. Their design is highly optimized and, what can be more important, reviewed.

**Comment by HarryR** [comment]
This has been implemented in #74 

I am reviewing the Zcash Sapling specification for further improvements. I think some of the improvements can't be used because of the parameters of the Baby JubJub curve (e.g. the optimisation where `a=-1`)
